### PR TITLE
Better support for CMake: TestBigEndian, CheckTypeSize, CMAKE_CROSSCOMPILING_EMULATOR

### PR DESCRIPTION
**Pull request by thewtex**
This overrides two CMake modules that are commonly used by projects, especially scientific projects or projects that aim to support a number of platforms.

The TestBigEndian modules is replaced.  This addresses:

  https://github.com/kripken/emscripten/issues/1106

Also the CheckTypeSize module is replaced.

  https://github.com/kripken/emscripten/issues/400

The fix for CheckTypeSize involves the use of the CMake try_run command.  Also ease the building of CMake project files by adding support for the CMAKE_CROSSCOMPILING_EMULATOR variable, and pointing it to nodejs.  This prevents the need to manually populate CMake TryRunResults.cmake files. It currently requires CMake Git master (to be released as  CMake 3.3).  Everything should still work with older CMake.

New tests are added to the test suite.  All tests passed when

  tests/runner.py

### CODE REVIEW

File: cmake/Modules/Platform/Emscripten.cmake [hunk 1]
```diff
@@ -258,3 +258,14 @@ if ("${CMAKE_GENERATOR}" MATCHES "^Visual Studio.*")
 	set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} ${EMSCRIPTEN_VS_LINKER_FLAGS}" CACHE STRING "")
 	set(CMAKE_MODULE_LINKER_FLAGS "${CMAKE_MODULE_LINKER_FLAGS} ${EMSCRIPTEN_VS_LINKER_FLAGS}" CACHE STRING "")
 endif()
+
+if (NOT DEFINED CMAKE_CROSSCOMPILING_EMULATOR)
+  find_program(node_js_executable NAMES node nodejs)
```
Comment (juj) [inline cmake/Modules/Platform/Emscripten.cmake]: Would this be better to be `NAMES nodejs node` to search for the more specific name first? See e.g. https://github.com/kripken/emscripten/issues/3296

**Comment by juj** [comment]
Thanks, this looks very good! One micro-comment above, but can merge either way. Let me know what you think.

**Comment by weitjong** [comment]
IMHO, there is an issue with this approach of overriding common CMake modules. For a multi-platform project, these overridden modules work only for Emscripten/HTML5 platform. This will force project maintainer to configure CMake module search path differently between Emscripten platform and other platforms. A better way is to use the original CMake modules and then enhance it with ```if(DEFINED ENV{EMSCRIPTEN})``` or something like that to decide which code path it should take. The original version or the "overridden" version. However, this makes the "enhanced" version of the CMake modules to be dependent on the future CMake upstream changes. But the proposed "overridden" CheckTypeSize module also suffer from this anyway. Just my two cents.

File: cmake/Modules/Platform/Emscripten.cmake (hunk 1 above)
Comment (thewtex) [inline cmake/Modules/Platform/Emscripten.cmake]: Yes, good point. I'll switch it.

**Comment by thewtex** [comment]
@weitjong Yes, eventually these fixes should be upstreamed into CMake.

**Comment by juj** [comment]
@weitjong: The current approach is to call `emcmake cmake`, which takes care of passing the correct command line parameters (including `-DCMAKE_MODULE_PATH`) correctly, so a project maintainer does not need to configure the module paths manually. This does have the limitation that a project cannot override `CMAKE_MODULE_PATH` with its own custom setting (or if it does, it has to be smart about when to do it, I think). If there is a good way around that, it would be good to know.

I don't think it's feasible to take the approach that Emscripten should not do "system" customizations to CMake, but that development should be done in upstream CMake, since especially in Linux world, people typically have very old CMakes in their systems, 2.8 seems to be quite popular, so it would be a very slow way to move forward. Having `emcmake` wrap cmake and set up `CMAKE_MODULE_PATH` has been the least painful path for quick iteration so far.

**Pull request merged**